### PR TITLE
Issue #2860 Fix leakage of HttpDestinations in HttpClient

### PR DESCRIPTION
**Pull request by AntonIvanov87**
Signed-off-by: Ivanov Anton <an.ivanov@corp.mail.ru>

### CODE REVIEW

**Comment by sbordet** [comment]
@AntonIvanov87 thanks for this PR.
Is there a reason you based it on 9.3.x? I would rather base your PR on 9.4.x.

**Comment by AntonIvanov87** [comment]
@sbordet, I just wanted newer 9.3.x versions to include this fix. If I base it on 9.4.x can what is the process of backporting it to 9.3.x?

**Comment by sbordet** [comment]
We can add the fix to 9.3.x but it's unlikely that we'll make a new 9.3.x release, as 9.4.x is the current Jetty release. I suggest that you move to 9.4.x if you can. Jetty 9.4.12 will be probably out next week.

Having said that, I'll accept your PR on 9.3.x and I will do the forward port to 9.4.x.

Thanks!

File: jetty-client/src/main/java/org/eclipse/jetty/client/PoolingHttpDestination.java [hunk 1]
```diff
@@ -210,16 +210,7 @@ public void close(Connection connection)
...
-            {
-                // There is a race condition between this thread removing the destination
-                // and another thread queueing a request to this same destination.
-                // If this destination is removed, but the request queued, a new connection
-                // will be opened, the exchange will be executed and eventually the connection
-                // will idle timeout and be closed. Meanwhile a new destination will be created
-                // in HttpClient and will be used for other requests.
-                getHttpClient().removeDestination(this);
-            }
+            removeDestinationIfIdle();
```
[the remark this answers is not in the document]
Comment (AntonIvanov87) [reply jetty-client/src/main/java/org/eclipse/jetty/client/PoolingHttpDestination.java]: Done.

File: jetty-client/src/main/java/org/eclipse/jetty/client/PoolingHttpDestination.java [hunk 2]
```diff
@@ -237,6 +228,27 @@ public void close()
         connectionPool.close();
     }
 
+    public void abort(Throwable cause)
+    {
+        super.abort(cause);
+        if (getHttpExchanges().isEmpty()) {
+            removeDestinationIfIdle();
+        }
```
[the remark this answers is not in the document]
Comment (AntonIvanov87) [reply jetty-client/src/main/java/org/eclipse/jetty/client/PoolingHttpDestination.java]: Sorry, have not got used to this style yet (

File: jetty-client/src/main/java/org/eclipse/jetty/client/PoolingHttpDestination.java [hunk 3]
```diff
@@ -237,6 +228,27 @@ public void close()
...
+    }
+
+    private void removeDestinationIfIdle() {
```
[the remark this answers is not in the document]
Comment (AntonIvanov87) [reply jetty-client/src/main/java/org/eclipse/jetty/client/PoolingHttpDestination.java]: Corrected, sorry.

File: jetty-client/src/test/java/org/eclipse/jetty/client/http/HttpDestinationOverHTTPTest.java [hunk 4]
```diff
@@ -280,6 +281,31 @@ public void testDestinationIsRemoved() throws Exception
         Assert.assertNotSame(destinationBefore, destinationAfter);
     }
 
+    @Test
+    public void testDestinationIsRemovedAfterConnectionError() throws Exception
+    {
+        String host = "localhost";
+        int port = connector.getLocalPort();
+        client.setRemoveIdleDestinations(true);
+        Assume.assumeTrue("Destinations of a fresh client must be empty", client.getDestinations().isEmpty());
```
[the remark this answers is not in the document]
Comment (AntonIvanov87) [reply jetty-client/src/test/java/org/eclipse/jetty/client/http/HttpDestinationOverHTTPTest.java]: Sure, corrected.

File: jetty-client/src/test/java/org/eclipse/jetty/client/http/HttpDestinationOverHTTPTest.java [hunk 5]
```diff
@@ -280,6 +281,31 @@ public void testDestinationIsRemoved() throws Exception
...
+
+        server.stop();
+        Request request = client.newRequest(host, port).scheme(this.scheme);
+        try
+        {
+            request.send();
+            Assume.assumeTrue("Request to a closed port must fail", false);
```
[the remark this answers is not in the document]
Comment (AntonIvanov87) [reply jetty-client/src/test/java/org/eclipse/jetty/client/http/HttpDestinationOverHTTPTest.java]: Sure.

File: jetty-client/src/test/java/org/eclipse/jetty/client/http/HttpDestinationOverHTTPTest.java [hunk 6]
```diff
@@ -280,6 +281,31 @@ public void testDestinationIsRemoved() throws Exception
...
+        } catch (Exception ignored) {
```
[the remark this answers is not in the document]
Comment (AntonIvanov87) [reply jetty-client/src/test/java/org/eclipse/jetty/client/http/HttpDestinationOverHTTPTest.java]: Ok, sorry.

File: jetty-client/src/test/java/org/eclipse/jetty/client/http/HttpDestinationOverHTTPTest.java [hunk 7]
```diff
@@ -280,6 +281,31 @@ public void testDestinationIsRemoved() throws Exception
...
+        }
+
+        long deadline = System.currentTimeMillis() + 100;
+        while (!client.getDestinations().isEmpty() && System.currentTimeMillis() < deadline)
+        {
+            Thread.yield();
+        }
```
[the remark this answers is not in the document]
Comment (AntonIvanov87) [reply jetty-client/src/test/java/org/eclipse/jetty/client/http/HttpDestinationOverHTTPTest.java]: I thought that a very short period of 100% cpu consumption is better than slowdown. But ok, changed to Thread.sleep(10).

**Comment by AntonIvanov87** [comment]
@sbordet, thank you for the review and recommendations.
I've also pushed the pull request for 9.4.x branch: https://github.com/eclipse/jetty.project/pull/2863

File: jetty-client/src/test/java/org/eclipse/jetty/client/http/HttpDestinationOverHTTPTest.java [hunk 8]
```diff
@@ -280,6 +280,32 @@ public void testDestinationIsRemoved() throws Exception
...
+        client.setRemoveIdleDestinations(true);
+        Assert.assertTrue("Destinations of a fresh client must be empty", client.getDestinations().isEmpty());
+
+        server.stop();
+        Request request = client.newRequest(host, port).scheme(this.scheme);
+        try
+        {
+            request.send();
+            Assert.fail("Request to a closed port must fail");
+        } catch (Exception expected)
```
[the remark this answers is not in the document]
Comment (AntonIvanov87) [reply jetty-client/src/test/java/org/eclipse/jetty/client/http/HttpDestinationOverHTTPTest.java]: Done

File: jetty-client/src/test/java/org/eclipse/jetty/client/http/HttpDestinationOverHTTPTest.java [hunk 9]
```diff
@@ -280,6 +280,32 @@ public void testDestinationIsRemoved() throws Exception
...
+        {
+        }
+
+        long deadline = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(100);
```
[the remark this answers is not in the document]
Comment (AntonIvanov87) [reply jetty-client/src/test/java/org/eclipse/jetty/client/http/HttpDestinationOverHTTPTest.java]: Done

**Comment by sbordet** [comment]
Thanks!

**Comment by AntonIvanov87** [comment]
Thanks)

**Pull request merged**